libcalico-go/lib/ipam/ipam_block_reader_writer.go

```diff
@@ -211,7 +211,11 @@ func (rw blockReaderWriter) claimAffineBlock(ctx context.Context, aff *model.KVP
 	logCtx := log.WithFields(log.Fields{"host": host, "subnet": subnet})
 
 	// Create the new block.
-	affinityKeyStr := "host:" + host
+	prefix := "host:"
+	if host == v3.VirtualLoadBalancer {
```

Comment:
I don't think this quite solves the problem - we're still effectively keying off the hostname in order to determine this.

So, someone can still make a Kubernetes node in their cluster that triggers this code.

I think we need to plumb this knowledge down through the IPAM code from the caller somehow, so that we can make sure that only the LB controller is ever generating affinities in this way.

Reply:
For example, we could:

- use the `IntendedUse` field of the `AutoAssignArgs` (and add it to `AssignIPArgs`)
- infer this from the provided `Attrs` map on `AutoAssignArgs` / `AssignIPArgs`
- add a new `AffinityType: Host | Virtual` option to those XArgs structs.

I think my preference is the first of those three (at least for now, while we only have a single scenario where it's possible for have a non-host affinity). 

If we ever end up making this more flexible, then option #3 might be better - but until then best not to enable combinations that we don't want.

Reply:
I think you will also need to look into these references to `"host:"` and see what the correct resolution is:

- https://github.com/caseydavenport/calico/blob/09903342c9c9927065afad5cbd5503947fbd29e7/libcalico-go/lib/ipam/ipam_block.go#L191-L194
- https://github.com/caseydavenport/calico/blob/09903342c9c9927065afad5cbd5503947fbd29e7/libcalico-go/lib/ipam/ipam_block.go#L196-L201
- https://github.com/caseydavenport/calico/blob/09903342c9c9927065afad5cbd5503947fbd29e7/libcalico-go/lib/ipam/ipam_block_reader_writer.go#L323-L330

Reply:
One thing to consider might be to replace all of the `host string` arguments passed around to something more generic - e.g.,

```go
type affinityConfig struct {
  type affinityType

  // The affinity value - a hostname for "host" type affinities. 
  value string
}
```

```go
func (c ipamClient) ensureBlock(ctx context.Context, rsvdAttr *HostReservedAttr, requestedPools []net.IPNet, version int, host string) (*net.IPNet, error) {
```
Would become

```go
func (c ipamClient) ensureBlock(ctx context.Context, rsvdAttr *HostReservedAttr, requestedPools []net.IPNet, version int, affinityCfg affinityConfig) (*net.IPNet, error) {
```

This would at least help ensure we're passing the necessary context wherever it might be needed.